service/storage_service.cc

```diff
@@ -3658,9 +3680,25 @@ future<> storage_service::handle_state_normal(inet_address endpoint, gms::permit
             do_remove_node(endpoint);
         }
     } else if (existing && *existing == endpoint) {
-        tmptr->del_replacing_endpoint(endpoint);
+        // This branch in taken by replace-with-different-ip.
+        // We actually don't remove anything from replacing endpoints here, since
+        // we pass new host_id to del_replacing_endpoint. The same was true
+        // with IP-based token_metadata - the new IP was passed to del_replacing_endpoint
+        // and it didn't remove anything. I'm not sure why we need this code at all.
```

Comment:
Not sure if I agree with / understand the comment.

Suppose node with host ID `id` and address `ip` was a replacing endpoint. So `tmptr` contains a replacing endpoint `id`.
Eventually replace operation finishes and it transitions to NORMAL state.
This triggers `handle_state_normal(ip)`, and we enter this branch because `existing == endpoint == ip`.

So no we need to remove `id` as a replacing endpoint, and make it normal instead.
It makes sense to me.

It also looks like we should remove the old ID as a "being_replaced endpoint" somewhere, but I don't see where this is happening.

Reply:
> It also looks like we should remove the old ID as a "being_replaced endpoint" somewhere, but I don't see where this is happening.

This is what the comment is talking about:
> // The old IP is removed by the next block of logic - we detach old IP from token ring,
> // it gets added to candidates_for_removal, then storage_service::remove_endpoint ->
> // _gossiper.remove_endpoint -> storage_service::on_remove -> it's removed
> // from topology and from replacing_endpoints.

Old node ("being_replaced") is removed because it's no longer referenced by tokens. `del_replacing_endpoint` is called as part of `token_metadata::remove_endpoint` <-`storage_service::on_remove`.

> Suppose node with host ID id and address ip was a replacing endpoint. So tmptr contains a replacing endpoint id.

In your scenario are we replacing with same of different IP?

For "replace-with-same-ip" we take the last branch, and we need to explicitly call `del_replacing_endpoint`
for old `host_id` because IP is now mapped to the new host_id and it won't be removed in the same way as for replace-with-different-ip above.

Reply:
> In your scenario are we replacing with same of different IP?

I was talking about replace with different IP, the same as the comment.

The replaced node is irrelevant in my reasoning.

The replacing node turns to a normal node at some point, and `handle_state_normal` is called for it. So I understand we need to call `del_replacing_endpoint` for the replacing node (i.e. new host ID) at this moment; it was a replacing endpoint, and now it no longer is a replacing endpoint, it is a normal endpoint.

Reply:
Ah, I think I'm just confused by the name `del_replacing_endpoint`.

The parameter it is given is not the ID of the replacing endpoint, but of the replaced endpoint...

Reply:
I propose to remove the branch. It's just confusing the reader.
We can do it in a follow-up.

Reply:
It might be leftover from times where host ID was being stolen during replace.
Then this branch would've been taken in the replace-with-same-IP scenario, and `del_replacing_endpoint(endpoint)` would make sense.

Reply:
> It might be leftover from times where host ID was being stolen during replace. Then this branch would've been taken in the replace-with-same-IP scenario, and `del_replacing_endpoint(endpoint)` would make sense.

I think so too